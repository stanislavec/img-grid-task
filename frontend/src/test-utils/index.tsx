import React, { ReactElement } from 'react';
import { createBrowserRouter, createMemoryRouter, RouterProvider } from 'react-router-dom';

export const withRouter = (Component: ReactElement<unknown, any>, path: string = '/') => {
  const router = createBrowserRouter([
    {
      path,
      element: Component,
    },
  ]);

  return <RouterProvider router={router} />;
};

export const withMemotyRouter = (
  Component: ReactElement<unknown, any>,
  { route = '/', path = '/' } = {},
) => {
  const router = createMemoryRouter(
    [
      {
        path,
        element: Component,
      },
    ],
    {
      initialEntries: [route],
    },
  );
  return <RouterProvider router={router} />;
};
