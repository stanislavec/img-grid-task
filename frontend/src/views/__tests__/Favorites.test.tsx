import { fireEvent, render, screen } from '@testing-library/react';
import usePhotoContext from 'context/useImageContext';
import { withRouter } from 'test-utils';
import Favorites from '../Favotites';
import PHOTOS from './__mocks__/photo';

jest.mock('context/useImageContext');

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockUseNavigate,
}));

const mockUsePhotoContext = usePhotoContext as jest.MockedFunction<typeof usePhotoContext>;

describe('Favorites', () => {
  beforeEach(() => jest.clearAllMocks());

  it('is defined', () => {
    mockUsePhotoContext.mockReturnValue({
      photos: [],
      fetchNext: () => null,
      loading: false,
      addToFavorite: () => null,
      removeFromFavorite: () => null,
      favorites: [],
    });
    const elem = render(withRouter(<Favorites />));
    expect(elem.baseElement).toBeDefined();
  });

  it('matches snapshot with photos', () => {
    mockUsePhotoContext.mockReturnValue({
      photos: PHOTOS,
      fetchNext: () => null,
      loading: false,
      addToFavorite: () => null,
      removeFromFavorite: () => null,
      favorites: PHOTOS.slice(0, 4),
    });
    const elem = render(withRouter(<Favorites />));
    expect(elem.container).toMatchSnapshot();
  });

  it('matches snapshot without photos', () => {
    mockUsePhotoContext.mockReturnValue({
      photos: [],
      fetchNext: () => null,
      loading: false,
      addToFavorite: () => null,
      removeFromFavorite: () => null,
      favorites: [],
    });
    const elem = render(withRouter(<Favorites />));
    expect(elem.container).toMatchSnapshot();
  });

  it('handles click on photo', () => {
    mockUsePhotoContext.mockReturnValue({
      photos: PHOTOS,
      fetchNext: () => null,
      loading: false,
      addToFavorite: () => null,
      removeFromFavorite: () => null,
      favorites: PHOTOS.slice(0, 4),
    });
    render(withRouter(<Favorites />));

    const image = screen.queryByAltText(PHOTOS[0].url);

    if (!image) throw Error();

    fireEvent.click(image);

    expect(mockUseNavigate).toBeCalledTimes(1);
  });
});
