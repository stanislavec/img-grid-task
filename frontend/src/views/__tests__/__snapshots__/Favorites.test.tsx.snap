// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites matches snapshot with photos 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/yC-Yzbqy7PY"
        src="https://picsum.photos/id/0/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/LNRyGwIJr5c"
        src="https://picsum.photos/id/1/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/N7XodRrbzS0"
        src="https://picsum.photos/id/2/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/Dl6jeyfihLk"
        src="https://picsum.photos/id/3/5000/3333"
      />
    </div>
  </div>
</div>
`;

exports[`Favorites matches snapshot without photos 1`] = `
<div>
  <div
    class="empty"
  >
    No data...
  </div>
</div>
`;
